features/jmx-config-generator/src/main/java/org/opennms/features/jmxconfiggenerator/jmxconfig/JmxDatacollectionConfiggenerator.java

```diff
@@ -110,266 +116,147 @@
-                                                     xmlMbean.getObjectname(),
-                                                     xmlJmxAttribute.getName()
-                                                             + " as "
-                                                             + xmlJmxAttribute.getAlias());
-                                        xmlMbean.getAttrib().add(xmlJmxAttribute);
-                                    }
-                                }
+        xmlJmxCollection.setName("JSR160-" + serviceName);
+        xmlJmxCollection.setRrd(rrd);
+        xmlJmxDatacollectionConfig.getJmxCollection().add(xmlJmxCollection);
+        xmlJmxCollection.setMbeans(new Mbeans());
+
+        Collection<FilterCriteria> attributeCriteriaCollection = Collections2.transform(ids, input -> FilterCriteria.parse(input));
+        for (QueryResult.MBeanResult eachResult : result.getMBeanResults()) {
+            ObjectName objectName = eachResult.objectName;
+
+            Mbean xmlMbean = new Mbean();
+            xmlMbean.setObjectname(objectName.toString());
+            String typeAndOthers = objectName.getCanonicalName().substring(objectName.getCanonicalName().lastIndexOf("=") + 1);
+            xmlMbean.setName(objectName.getDomain() + "." + typeAndOthers);
+
+            MBeanInfo jmxMbeanInfo = eachResult.mbeanInfo;
+            for (MBeanAttributeInfo jmxBeanAttributeInfo : jmxMbeanInfo.getAttributes()) {
+                for (FilterCriteria eachCriteria : attributeCriteriaCollection) {
+                    if (eachCriteria.matches(jmxBeanAttributeInfo)) {
+                        // check for CompositeData
+                        if ("javax.management.openmbean.CompositeData".equals(jmxBeanAttributeInfo.getType())) {
+                            CompAttrib compAttrib = createCompAttrib(mBeanServerConnection, objectName, jmxBeanAttributeInfo);
+                            if (compAttrib != null) {
+                                xmlMbean.getCompAttrib().add(compAttrib);
                             }
                         }
-
-                        if (xmlMbean.getAttrib().size() > 0
-                                || xmlMbean.getCompAttrib().size() > 0) {
-                            xmlJmxCollection.getMbeans().getMbean().add(xmlMbean);
+                        if (numbers.contains(jmxBeanAttributeInfo.getType())) {
+                            Attrib xmlJmxAttribute = createAttr(jmxBeanAttributeInfo);
+                            xmlMbean.getAttrib().add(xmlJmxAttribute);
                         } else {
-                            logger.debug("mbean: " + xmlMbean.getName()
-                                    + " has no relavant attributes.");
+                            logger.warn("The type of attribute '{}' is '{}', but only numbers are supported right now. Ignoring.", jmxBeanAttributeInfo.getName(), jmxBeanAttributeInfo.getType());
                         }
                     }
```

Comment:
Is it possible that an attribute will be added for each matching filter criteria even it is already adde to the xmlBean?